contracts/libraries/FVMLib.sol

```diff
@@ -569,5 +632,14 @@ function decodeSwap(bytes calldata data)
                 calldataload(add(data.offset, add(1, pointer)))
             )
         output := mul(base1, exp(10, power1))
+
+        // Checks if one of these variables overflows
+        if or(
+            gt(input, 0xffffffffffffffffffffffffffffffff),
+            gt(output, 0xffffffffffffffffffffffffffffffff)
```

Comment:
the overflow checks don't work like this because the overflow can already have happened in `output := mul(base1, exp(10, power1))`.  `base1 = 1, power1 = 78` => `mul(1, 1e78)` overflows 256 bits and would result in the value mod 2**256. Maybe you should just do the decoding in assembly and the final multiplication for `input` and `output` in Solidity which defaults to multiplication with overflow checks.

Reply:
Same issue at the other decoding functions

Reply:
Or if you want to stay in assembly you could do the reverse check in addition to checking if it's < 2**128:

```
p1 := exp(10, power1)
output := mul(base1, p1)

// something like that should work iff output did not overflow
eq(base1, div(output, p1)
```

Reply:
Thanks for pointing that out, I missed that part. I just pushed a fix for the `decodeSwap` function, if it's correct I'll update the rest of the decoding functions.

Reply:
The updated version looks correct to me. I'll note that the operations to determine `length0` and `length1` could have underflow as well, but I think the most likely consequence of that is both swap values being zero (`base0` and `base1` will get right shifted out of existence) so shouldn't be a big risk to caller that messes up their calldata.